### ISSUE TITLE
Start orchestration cross function app

### ISSUE THREAD

**Issue by mwikstrom** [Azure/azure-functions-durable-extension#1281]
**Is your feature request related to a problem? Please describe.**
We want to separate orchestration logic and web api logic into different function apps.

In particular, we want http triggers in the api app to query, start and terminate orchestrations.

We assumed that this would be a possible setup if we ensured that both apps were configured with the same task hub name.

However, when attempting to start an orchestration from the api app an exception is thrown with the following message:

>The function 'BuildOrchestration' doesn't exist, is disabled, or is not an orchestrator function. Additional info: No orchestrator functions are currently registered!

There is however an orchestrator function named 'BuildOrchestration' in the other function app, and we believe that if this check would be ignored then things would work as we want since interaction between the two apps goes via the common task hub.

**Describe the solution you'd like**
I would like to have an option that skips the check whether a started orchestration is mapped to a function in the currrent app, or even better to be able to register "external orchestrator functions".

**Describe alternatives you've considered**
We're currently using the following hack:

```c#
/// <summary>
///     Starts an orchestration that is hosted by another function app, using the same task hub name.
/// </summary>
/// <remarks>
///     This is a reflection hack that register the specified orchestrator function name in the current app
///     just before calling `StartNewAsync(...)`.
///     
///     IMPORTANT! Do NOT use this method for functions that DO exist in the current app, doing so will
///     overwrite their registration causing undefined behavior.
/// </remarks>
public static Task StartNewExternalAsync<TInput>(
    this IDurableOrchestrationClient client,
    string orchestratorFunctionName,
    string instanceId,
    TInput input)
{
    Requires.ArgNotNull(client, nameof(client));
    Requires.ArgNotNull(orchestratorFunctionName, nameof(orchestratorFunctionName));
    Requires.ArgNotNull(instanceId, nameof(instanceId));

    var clientType = client.GetType();
    var configField = clientType.GetField("config", BindingFlags.NonPublic | BindingFlags.Instance);

    if (configField is null || configField.FieldType != typeof(DurableTaskExtension))
    {
        throw new NotSupportedException($"Cannot extract '{nameof(DurableTaskExtension)}' from '{clientType}'");
    }

    var config = (DurableTaskExtension)configField.GetValue(client)!;
            
    var registerMethod = typeof(DurableTaskExtension).GetMethod(
        "RegisterOrchestrator",
        BindingFlags.NonPublic | BindingFlags.Instance);

    if (registerMethod is null)
    {
        throw new NotSupportedException(
            $"Unable to find the internal 'RegisterOrchestrator' method in '{nameof(DurableTaskExtension)}'");
    }

    var functionNameType = typeof(DurableTaskExtension).Assembly.GetType(
        "Microsoft.Azure.WebJobs.Extensions.DurableTask.FunctionName");

    if (functionNameType is null)
    {
        throw new NotSupportedException("Unable to find required internal 'FunctionName' type");
    }

    var functionNameParam = Activator.CreateInstance(functionNameType, orchestratorFunctionName);

    registerMethod.Invoke(config, new[] { functionNameParam, null });

    return client.StartNewAsync(orchestratorFunctionName, instanceId, input);
}
```

With this hack we're able to start orchestrations that are hosted by the other app.

**Comment by cgillum** [comment]
FYI @amdeel this looks similar to the signal-entities issue you fixed [here](https://github.com/Azure/azure-functions-durable-extension/pull/1262). Could we take the same approach for orchestration methods, like `StartNewAsync`?

**Comment by olitomlinson** [comment]
FYI When I was working on another issue, I once made a custom build of the extension and removed this check here :

https://github.com/Azure/azure-functions-durable-extension/blob/f2e92b866aed1d085eed33d2a7bf50d9aecfb70d/src/WebJobs.Extensions.DurableTask/DurableOrchestrationClient.cs#L81

This allowed me start an orchestration from a different Function App than where the orchestration was hosted.

Suggestion, could we have an additional parameter on `StartNewAsync` that skips the check?

Something like :

```
await client.StartNewAsync(UseExternalOrchestration: true)
```

Could this approach cause problems when upgrading to a newer version of the DF extension? I imagine both function Apps would have to be shut down and both upgraded at the same time, otherwise we could run into compatibility issues?

**Comment by ConnorMcMahon** [comment]
Is there a reason we do this check when TaskHub or ConnectionName are provided? It seems like if either of those is provided, we are likely to be communicating cross app, and shouldn't be checking that we have the function we are trying to call.

**Comment by mwikstrom** [comment]
@ConnorMcMahon I believe that HubName is sometimes set just to customize storage naming and therefore it might be a bad idea to skip this check based on having that setting.

I would prefer an approach where external functions are "registered". 

This could be done in a setting like this:

```json
{
  "version": "2.0",
  "extensions": {
    "durableTask": {
      "hubName": "myhub",
      "externalOrchestratorFunctions": [ "FunctionA", "FunctionB" ]
    }
  }
}
```

I would also like to have a special value for declaring that all orchestrator functions are external:

```json
{
  "version": "2.0",
  "extensions": {
    "durableTask": {
      "hubName": "myhub",
      "externalOrchestratorFunctions": true
    }
  }
}
```

With this notion, the default value would be `false` indicating that no orchestrator functions are external (the current situation).

**Comment by ConnorMcMahon** [comment]
@mwikstrom, I should have made myself more clear in my last mesesage.

I was not referring to setting `hubName` or `connectionName` in the host.json. I was referring to the fields `taskHub` and `connectionName` on the durable client [binding](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-bindings#orchestration-client). By default, these fields will use the default values used for the TaskHub + Storage Account used by this application. By specifying something else in the binding configuration, it seems to be indicating that the client will be used for cross-app communication, signaling we shouldn't be checking that the function we are trying to execute exists on our application.

**Comment by amdeel** [comment]
Enabling/fixing this will be done in two steps. First, I will only call ThrowIfFunctionDoesNotExist if we are using the same connectionName and taskHub in the DurableClient as we have stored in our configuration settings. This is exactly like the work  done for this issue: https://github.com/Azure/azure-functions-durable-extension/issues/1248

Then, I will add another method overload to allow connectionName and taskHub to be passed in as parameters, similarly to how we currently do SignalEntityAsync. Because this is something of a breaking change, I will only do this second step for our 2.3.0 release.

After I merge my PR for this first step, you will be able to start orchestrations in other functions apps when you specify a ConnectionName and TaskHub in the DurableClient attribute.

**Comment by cgillum** [comment]
The PR for this was merged and the fix will go out with the v2.3.0 release.

**Comment by amdeel** [comment]
Work adding the taskHub and connectionName parameters to methods to IDurableOrchestrationClient methods has been scrapped.

The previous work for this is in DurableFunctions starting 2.2.1, as stated above.
